packages/core/src/extensions/Blocks/api/serialization.ts

```diff
@@ -1,29 +1,114 @@
 import { Extension } from "@tiptap/core";
 import { Plugin } from "prosemirror-state";
-import { DOMSerializer, Schema } from "prosemirror-model";
-
-const customBlockSerializer = (schema: Schema) => {
-  const defaultSerializer = DOMSerializer.fromSchema(schema);
-
-  return new DOMSerializer(
-    {
-      ...defaultSerializer.nodes,
-      // TODO: If a serializer is defined in the config for a custom block, it
-      //  should be added here. We still need to figure out how the serializer
-      //  should be defined in the custom blocks API though, and implement that,
-      //  before we can do this.
-    },
-    defaultSerializer.marks
-  );
+import { DOMSerializer, Fragment, Node, Schema } from "prosemirror-model";
+import { nodeToBlock } from "../../../api/nodeConversions/nodeConversions";
+import { BlockNoteEditor } from "../../../BlockNoteEditor";
+import { BlockSchema, SpecificBlock } from "./blockTypes";
+
+function doc(options: { document?: Document }) {
+  return options.document || window.document;
+}
+
+export const customBlockSerializer = <BSchema extends BlockSchema>(
+  schema: Schema,
+  editor: BlockNoteEditor<BSchema>
+) => {
+  const customSerializer = DOMSerializer.fromSchema(schema) as DOMSerializer & {
+    serializeNodeInner: (
+      node: Node,
+      options: { document?: Document }
+    ) => HTMLElement;
+  };
+
+  customSerializer.serializeNodeInner = (
```

Comment:
can we extend DomSerializer (create a new class), instead of overwriting a method like this? Not sure if it's possible in this case

Reply:
Don't really even need to extend it, the constructor for `DOMSerializer` allows you to specify serialization rules for different node types, so I can just set one for `blockContainer`

Reply:
I tried 2 different methods of refactoring the `DOMSerializer`. First was using the existing `DOMSerializer` constructor and adding a custom serialization rule for `BlockContainer` (see [commit b4f3fb6](https://github.com/TypeCellOS/BlockNote/pull/257/commits/b4f3fb6433c6a9e00b90a4f3a6f7d1ff87bc1758)), and the second by extending the DOMSerializer class (see the [custom-block-serialization-class-extension](https://github.com/TypeCellOS/BlockNote/tree/custom-block-serialization-class-extension) branch).

While both of these work for copy/paste, for some reason the serializer still tries to read `toDOM` for custom blocks in other cases like converting to Markdown & HTML. I don't understand why this is the case, especially since AFAIK the second method should literally do the same thing as overwriting `serializeNodeInner` like how it's done on this branch. Maybe you could have a look and see if there's anything obvious that's wrong?